src/blueapi/cli/scratch.py

```diff
@@ -128,3 +131,95 @@ def _validate_directory(path: Path) -> None:
         raise KeyError(f"{path}: No such file or directory")
     elif path.is_file():
         raise KeyError(f"{path}: Is a file, not a directory")
+
+
+def _get_project_name_from_pyproject(path: Path) -> str:
+    pyproject_path = path / "pyproject.toml"
+    if pyproject_path.exists():
+        with pyproject_path.open("r", encoding="utf-8") as file:
+            toml_data = parse(file.read())
+        return toml_data.get("project", {}).get("name", "")
+    return ""
+
+
+def _fetch_installed_packages_details() -> list[PackageInfo]:
+    installed_packages = importlib.metadata.distributions()
+    return [
+        PackageInfo(
+            name=dist.metadata["Name"],
+            version=dist.version,
+            location=str(dist.locate_file("")),
+            is_dirty=False,
+        )
+        for dist in installed_packages
+    ]
+
+
+def get_python_environment(
+    config: ScratchConfig | None,
+    name: str | None = None,
+    source: SourceInfo | None = None,
+) -> PythonEnvironmentResponse:
+    """
+    Get the Python environment. This includes all installed packages and
+    the scratch packages.
+    """
+    scratch_packages = {}
+    packages = []
+
+    if config is None:
+        python_env_response = PythonEnvironmentResponse(scratch_enabled=False)
+    else:
+        python_env_response = PythonEnvironmentResponse(scratch_enabled=True)
+        try:
+            _validate_directory(config.root)
+        except Exception as e:
+            logging.error(f"Failed to get scratch info: {e}")
+            return python_env_response
```

Comment:
nit: If there's an exception, the caller [this method is called from the rest `interface.py`] won't be informed if they don't check the logs.

I think we should allow the Exception to escape here, and return a non-200 status code.